tests/nix/default.nix

```
@@ -52,78 +45,136 @@ in
+          # - https://discord.com/channels/1209971237770498088/1262564341413056632/1360866631214956615
+          # - https://nix.dev/manual/nix/latest/release-notes/rl-2.26 (possible breaking changes here)
+
+          # "nix"
+          "lix"
+        ];
+        testName = builtins.attrNames cfg;
+      };
+      mkTest =
+        {
+          testName,
+          opts,
+          nix,
+          nixpkgs,
+        }:
+        let
+          sanitizeName =
+            str: lib.strings.sanitizeDerivationName (builtins.replaceStrings [ "." ] [ "_" ] str);
+          identifier = sanitizeName "${nixpkgs.legacyPackages.lib.version}-${
+            nixpkgs.legacyPackages.${nix}.name
+          }";
+          path = "tests/nix/suite/${testName}";
+          injectedFlakeDir = pkgs.runCommand "injected-flake-dir" { } ''
+            cp -r ${../..} $out
+            chmod -R +w $out
+            substituteInPlace $out/${path}/hive.nix --replace @IDENT@ ${identifier}
+          '';
+        in
+        rec {
+          name = "nixos-vm-test-${testName}-${identifier}";
+          value = nixpkgs.legacyPackages.testers.runNixOSTest {
+            inherit (opts) nodes;
+            inherit name;
+            defaults =
+              {
+                pkgs,
+                evaluateHive,
+                ...
+              }:
+              let
+                hive = evaluateHive {
+                  nixpkgs = pkgs.path;
+                  path = injectedFlakeDir;
+                  hive = builtins.scopedImport {
+                    __nixPath = _b: null;
+                    __findFile = path: name: if name == "nixpkgs" then pkgs.path else throw "oops!!";
+                  } "${injectedFlakeDir}/${path}/hive.nix";
+                };
+                nodes = mapAttrsToList (_: val: val.config.system.build.toplevel.drvPath) hive.nodes;
+                # fetch **all** dependencies of a flake
+                # it's called fetchLayer because my naming skills are awful
+                fetchLayer =
+                  input:
+                  let
+                    subLayers = if input ? inputs then map fetchLayer (builtins.attrValues input.inputs) else [ ];
+                  in
+                  [
+                    input.outPath
+                  ]
+                  ++ subLayers;
+              in
+              {
+                imports = [ ./test-opts.nix ];
+                nix = {
+                  package = pkgs.${nix};
+                  nixPath = [ "nixpkgs=${pkgs.path}" ];
+                  settings.substituters = lib.mkForce [ ];
+                };
 
-              virtualisation.memorySize = 4096;
-              virtualisation.additionalPaths = flatten (nodes ++ (mapAttrsToList (_: fetchLayer) inputs));
+                virtualisation.memorySize = 4096;
+                virtualisation.additionalPaths = flatten (
+                  [ injectedFlakeDir ] ++ nodes ++ (mapAttrsToList (_: fetchLayer) inputs)
+                );
```

```suggestion
                virtualisation.additionalPaths = flatten [
                    injectedFlakeDir
                    nodes
                    (mapAttrsToList (_: fetchLayer) inputs)
                ];
```

this is something I forgot to fix but it does look a lot nicer.